cli/azd/internal/telemetry/storage.go

```diff
@@ -0,0 +1,277 @@
+// Once the item is processed, consumers are responsible for calling Remove to remove the item from the queue.
+type StorageQueue struct {
+	folder            string
+	itemFileExtension string
+
+	// Standard time library clock, unless mocked in tests
+	clock clock.Clock
+}
+
+type StoredItem struct {
+	// Number of retries attempted
+	retryCount int
+
+	// Message in the item
+	message []byte
+
+	// File name of the stored item
+	fileName string
+}
+
+func (itm *StoredItem) RetryCount() int {
+	return itm.retryCount
+}
+
+func (itm *StoredItem) Message() []byte {
+	return itm.message
+}
+
+type itemEntry struct {
+	name        string
+	readyTime   time.Time
+	fileModTime time.Time
+	retryCount  int
+}
+
+// Creates the storage-based queue.
+func NewStorageQueue(folder string, itemFileExtension string) (*StorageQueue, error) {
+	if err := os.MkdirAll(folder, osutil.PermissionDirectory); err != nil {
+		return nil, fmt.Errorf("failed to create storage queue folder: %v", err)
+	}
+
+	if !strings.HasPrefix(itemFileExtension, ".") {
+		itemFileExtension = "." + itemFileExtension
+	}
+
+	storage := StorageQueue{
+		folder:            folder,
+		itemFileExtension: itemFileExtension,
+		clock:             clock.New(),
+	}
+	return &storage, nil
+}
+
+// Queues a message.
+func (stg *StorageQueue) Enqueue(message []byte) error {
+	return stg.save(time.Duration(0), 0, message)
+}
+
+// Queues a message with delay.
+func (stg *StorageQueue) EnqueueWithDelay(message []byte, delayDuration time.Duration, retryCount int) error {
+	return stg.save(delayDuration, retryCount, message)
+}
+
+func (stg *StorageQueue) save(delayDuration time.Duration, retryCount int, message []byte) error {
+	file, err := os.CreateTemp(stg.folder, "*_itm.tmp")
+	if err != nil {
+		return fmt.Errorf("failed to create temp file :%v", err)
+	}
+
+	err = os.WriteFile(file.Name(), message, osutil.PermissionFile)
+	if err != nil {
+		return fmt.Errorf("failed to write file: %v", err)
+	}
+	file.Close()
+
+	generatedFileName := filepath.Base(file.Name())
+	randomSuffix := generatedFileName[:strings.LastIndex(generatedFileName, "_")]
+	readyTime := stg.clock.Now().Add(delayDuration)
+	fileName := formatFileName(readyTime, retryCount, randomSuffix, stg.itemFileExtension)
+	err = os.Rename(file.Name(), filepath.Join(stg.folder, fileName))
+	if err != nil {
+		return fmt.Errorf("failed to rename file: %v", err)
+	}
+
+	return nil
+}
+
+// Gets the next available item for processing.
+// Returns nil if no items exist.
+// Returns error if an error occurs while reading storage.
+func (stg *StorageQueue) Peek() (*StoredItem, error) {
+	items, err := stg.getAllItems()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get stored files: %v", err)
+	}
+
+	latestTime := time.Time{}
+	latestIndex := -1
+	now := stg.clock.Now()
+	for i, item := range items {
+		if now.Sub(item.readyTime) >= 0 && item.fileModTime.After(latestTime) {
```

Comment:
Am I missing something, or is this LIFO policy? Comments at the beginning suggest a FIFO queue?

Reply:
It is supposed to be a FIFO. A latest refactor that I had split out `readyTime` vs `fileModTime` had reversed the behavior :). Will address